src/sqlfluff/rules/L003.py

```diff
@@ -318,96 +407,57 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
-                "trigger": None,
-            }
-        else:
-            memory = context.memory
-
-        if context.segment.is_type("newline"):
-            memory["in_indent"] = True
-        elif memory["in_indent"]:
-            if context.segment.is_type("whitespace"):
-                # it's whitespace, carry on
-                pass
-            elif context.segment.segments or (
-                context.segment.is_meta
-                and context.segment.indent_val != 0  # type: ignore
-            ):
-                # it's not a raw segment or placeholder. Carry on.
-                pass
-            else:
-                memory["in_indent"] = False
-                # we're found a non-whitespace element. This is our trigger,
-                # which we'll handle after this if-statement
-                memory["trigger"] = context.segment
-        else:
-            # Not in indent and not a newline, don't trigger here.
-            pass
-
-        # Is this the last segment? If so, need to "flush" any leftovers.
-        is_last = self._is_last_segment(
-            context.segment, memory, context.parent_stack, context.siblings_post
+        is_ignorable = any(
+            el.is_type(*self._ignore_types) for el in context.parent_stack + (segment,)
         )
+        if is_ignorable:
+            return LintResult(memory=memory)
 
-        if not context.segment.is_type("newline") and not is_last:
-            # We only process complete lines or on the very last segment
-            # (since there may not be a newline on the very last line)..
+        if segment.is_type("newline"):
+            memory.in_indent = True
+        elif memory.in_indent:
+            is_non_raw_segment = bool(segment.segments)
+            is_placeholder = segment.is_meta and segment.indent_val != 0  # type: ignore
+            if not (segment.is_whitespace or is_non_raw_segment or is_placeholder):
+                memory.in_indent = False
+                # First non-whitespace element is our trigger
+                memory.trigger = segment
+
+        is_last = self.is_final_segment(context)
+        if not segment.is_type("newline") and not is_last:
+            # Process on line ends or file end
             return LintResult(memory=memory)
 
-        if raw_stack and raw_stack[-1] is not context.segment:
-            raw_stack = raw_stack + (context.segment,)
-        res = self._process_raw_stack(
-            raw_stack,
-            memory,
+        line_summaries = self._process_raw_stack(
+            raw_stack=raw_stack,
+            memory=memory,
             tab_space_size=self.tab_space_size,
             templated_file=context.templated_file,
         )
+        memory.line_summaries = line_summaries
+        trigger_segment = memory.trigger
+        memory.trigger = None
+        if line_summaries and trigger_segment:
+            last_line_no = max(line_summaries.keys())
+            this_line = line_summaries[last_line_no]
```

Comment:
@barrywhart What say you to a Small data class for the Dict of _LineSummaries that knows a few of these kinds of get methods

Reply:
Sounds great to me. I've been adding lots of data classes, named tuples, and type annotations the past year or two, especially after we stopped supporting Python 2.7.

Reply:
Yea it's more a question of if the methods would be "too involved" I think it's safer to keep data inert (this code already suffers heavily from being mutative and being confused about what is "doing" and what is just data)